### PR TITLE
Add objects GC in dataset iterator

### LINKED ISSUE #33846: [data] release test failure : pipelined_ingestion_1500_gb
### What happened + What you expected to happen

looks like it timed out 

[ERROR 2023-03-28 19:02:48,645] run_release_test.py: 164  Command timed out after 9600.080104424998 seconds.
--
  | Traceback (most recent call last):
  | File "ray_release/scripts/run_release_test.py", line 160, in main
  | no_terminate=no_terminate,
  | File "/tmp/release-a8qe7WenMw/release/ray_release/glue.py", line 488, in run_release_test
  | raise pipeline_exception
  | File "/tmp/release-a8qe7WenMw/release/ray_release/glue.py", line 372, in run_release_test
  | raise e
  | File "/tmp/release-a8qe7WenMw/release/ray_release/glue.py", line 364, in run_release_test
  | raise_on_timeout=not is_long_running,
  | File "/tmp/release-a8qe7WenMw/release/ray_release/command_runner/anyscale_job_runner.py", line 269, in run_command
  | job_status_code, error, raise_on_timeout=raise_on_timeout
  | File "/tmp/release-a8qe7WenMw/release/ray_release/command_runner/anyscale_job_runner.py", line 174, in _handle_command_output
  | f"Command timed out after {workload_time_taken} seconds."
  | ray_release.exception.TestCommandTimeout: Command timed out after 9600.080104424998 seconds.

### Versions / Dependencies

master

### Reproduction script

https://buildkite.com/ray-project/release-tests-branch/builds/1493#01872a7e-b392-4a66-98ea-21091dc3636f

### Issue Severity

None

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
Fix: https://github.com/ray-project/ray/issues/33846

The DatasetIterator doesn't eagerly GC objects, which resulted in OOM of consumer nodes. The new consumer nodes that got brought up were not in sync with other healthy consumer nodes. The DatasetPipeline requires all consumers to read windows in sync, so this caused the pipeline to hang and then fail with timeout.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_object_gc.py [hunk 1]
```diff
@@ -24,6 +35,8 @@ def test_iter_batches_no_spilling_upon_no_transformation(shutdown_only):
 
     check_no_spill(ctx, ds.repeat())
     check_no_spill(ctx, ds.window(blocks_per_window=20))
+    check_to_torch_no_spill(ctx, ds.repeat())
```
Comment (jianoaix) [inline python/ray/data/tests/test_object_gc.py:38]: This will fail (i.e. have spilling) if without this PR.

**Review by jianoaix — COMMENTED** [review]
Tested: https://buildkite.com/ray-project/release-tests-pr/builds/33701#0187499d-df5a-4e4b-adac-fdb0bc32dca9

**Comment by ericl** [comment]
Nice find!

**Review by ericl — APPROVED** [review]

File: python/ray/data/_internal/dataset_iterator/dataset_iterator_impl.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def _to_block_iterator(
         ds = self._base_dataset
         block_iterator, stats, executor = ds._plan.execute_to_iterator()
         ds._current_executor = executor
-        return block_iterator, stats
+        return block_iterator, stats, False
```
Comment (c21) [inline python/ray/data/_internal/dataset_iterator/dataset_iterator_impl.py:33]: shall we update the type hint at line 28 as well?

File: python/ray/data/_internal/dataset_iterator/pipelined_dataset_iterator.py [hunk 3]
```diff
@@ -35,11 +35,20 @@ def _to_block_iterator(
...
         epoch_pipeline = self._get_next_dataset()
 
+        if epoch_pipeline._first_dataset is not None:
+            blocks_owned_by_consumer = (
+                epoch_pipeline._first_dataset._plan.execute()._owned_by_consumer
+            )
+        else:
+            blocks_owned_by_consumer = (
+                epoch_pipeline._peek()._plan.execute()._owned_by_consumer
+            )
```
Comment (c21) [inline python/ray/data/_internal/dataset_iterator/pipelined_dataset_iterator.py:45]: could you add a comment in code for why we need to do this?

**Review by c21 — APPROVED** [review]

File: python/ray/data/_internal/dataset_iterator/pipelined_dataset_iterator.py (hunk 3 above)
Comment (jianoaix) [reply python/ray/data/_internal/dataset_iterator/pipelined_dataset_iterator.py:49]: Done.

**Review by jianoaix — COMMENTED** [review]
The pipelined_ingestion_1500_gb has been consistently passing: https://buildkite.com/ray-project/release-tests-pr/builds?branch=jianoaix%3Aiteratorgcblocks
Will wait the CI to pass and then merge.

**Comment by jianoaix** [comment]
There is failure in python/ray/data/tests/test_dataset_consumption.py:: test_dataset_lineage_serialization_unsupported, but it's not relevant here.

**Pull request merged**